onnxruntime/contrib_ops/webgpu/bert/flash_attention.cc

```diff
@@ -26,73 +26,79 @@ Status CopyKVCacheProgram::GenerateShaderCode(ShaderHelper& shader) const {
   // This makes it so that FlashAttention only needs to look at present key and value, and saves
   // number of input buffers in the shader, which we run out of (<=8) without this optimization.
   const int components = parameters.head_size_ % 4 == 0 ? 4 : (parameters.head_size_ % 2 == 0 ? 2 : 1);
-  bool has_past = (past_sequence_length != 0);
-  CopyKVCacheProgram program{"CopyKVCache", has_past};
-  if (has_past) {
-    program.AddInputs({{K, ProgramTensorMetadataDependency::TypeAndRank, components},
-                       {V, ProgramTensorMetadataDependency::TypeAndRank, components},
-                       {past_key, ProgramTensorMetadataDependency::TypeAndRank, components},
-                       {past_value, ProgramTensorMetadataDependency::TypeAndRank, components}});
-  } else {
+  TensorShape valid_present_shape{parameters.batch_size_, parameters.num_heads_, parameters.total_sequence_length_, parameters.head_size_ / components};
+  int64_t valid_kv_size = valid_present_shape.Size();
+  bool has_past = (parameters.past_sequence_length_ > 0);
+  CopyKVCacheProgram program{"CopyKVCache", has_past, parameters.qkv_format_ == Q_K_V_BSNH_BNSH_BNSH, parameters.past_present_share_buffer_};
+  if (parameters.qkv_format_ == Q_K_V_BSNH_BNSH_BNSH) {
     program.AddInputs({{K, ProgramTensorMetadataDependency::TypeAndRank, components},
                        {V, ProgramTensorMetadataDependency::TypeAndRank, components}});
+  } else {
+    TensorShape reshaped_KV_shape{parameters.batch_size_, parameters.kv_sequence_length_, parameters.num_heads_, parameters.head_size_ / components};
+    program.AddInputs({{K, ProgramTensorMetadataDependency::TypeAndRank, reshaped_KV_shape, components},
+                       {V, ProgramTensorMetadataDependency::TypeAndRank, reshaped_KV_shape, components}});
+  }
+  if (has_past && !parameters.past_present_share_buffer_) {
+    program.AddInputs({{past_key, ProgramTensorMetadataDependency::TypeAndRank, components},
+                       {past_value, ProgramTensorMetadataDependency::TypeAndRank, components}});
   }
-
   program.AddOutputs({{present_key, ProgramTensorMetadataDependency::Rank, components},
-                      {present_value, ProgramTensorMetadataDependency::Rank, components}});
-
-  program.SetDispatchGroupSize(total_sequence_length, 1, parameters.num_heads_)
-      .SetWorkgroupSize(1)
-      .CacheHint(std::to_string(components) + std::to_string(has_past))
-      .AddUniformVariables({{static_cast<uint32_t>(past_sequence_length)},
-                            {static_cast<uint32_t>(parameters.kv_sequence_length_)},
-                            {static_cast<uint32_t>(parameters.head_size_ / components)}});
+                      {present_value, ProgramTensorMetadataDependency::Rank, components}})
+      .AddIndices(valid_present_shape);
+  program.SetDispatchGroupSize(gsl::narrow<uint32_t>(valid_kv_size + 63 / 64))
```

Comment:
nit: jFYI, this is good for now - but once we have generation mode FA, we only need to dispatch for the size of the new entries in the KV cache. uniforms.valid_present_size will also change to be uniforms.valid_new_kv_size

Reply:
The data is stored as [num_heads, total_sequence_length, head_size]. If we insert one new sequence in head 0 to the KV cache. The following data needs to be shifted backward. We can't keep the existed data's position unchanged. So we still need to dispatch the valid present size. Am I correct?

Reply:
Yes you are correct except when we have static kv cache which is supported for gqa, then all we need to do is copy in the new kv data

Reply:
Got it. But currently, even for static kv cache, we still need to copy the whole buffer. The only advantage of static kv cache is that we avoid dynamically create buffers since we allocated a big buffer in advance.

Reply:
Sorry, my misunderstanding. For static kv cache, the present_sequence_length is always max sequence length. So we only need to copy the new kv data like you said for generation mode.